### PR TITLE
Processing replication guides level by level

### PR DESCRIPTION
**Pull request by ke-yu**
### Purpose

This PR updates replication guide processing. 

Previously, replication guide processing is based on the numeric value of replication guide number. For example, in `foo(xs<1>, ys<2><3>, zs<3>)`, replication guide numbers are processed in ascendant order one by one: 1 -> 2 -> 3. For replication guide number that appears in multiple locations, zip replication will be applied to; otherwise Cartesian replication will be applied to, so the execution engine will generate the following replications:
  * Cartesian replication on xs
  * Cartesian replication on ys
  * Zip replication on ys and zs

Though this approach works for most cases, it has a prerequisite that replication guides for each arguments should be in ascendant order as well. For example, replication guides in `foo(xs<3><1>, ys<4><2>)` are considered as invalid.

As there is no "level" concept in replication guide processing, `foo(xs<1><1>)` is meaningless because we cannot zip on the same argument. It also has limitation like for two arguments, how to do zip replication on the first level and Cartesian replication on the second level. 

This PR changes replication guides processing. Now the processing is done level by level. For example, for the following function call, there are two levels for replication guides:

untitled

For each replication level - 
* Sort replication guides in ascendant order
* For each replication guide number
  - If it appears in multiple locations, zip replication will be applied to. By default using shortest lacing. If any replication guide number has suffix “L”, longest lacing applies.
  - Otherwise Cartesian replication will be applied to.

So for this example, the following replications will be applied to the arguments:
* Zip replication on `as`, `cs`
* Cartesian replication on `ds`
* Zip replication on `as`, `ds`
* Cartesian replication on `cs`

Now for `foo(xs<1><1>)`, its meaning is zooming into the third level of `xs` and call function `foo()`. For example, 
```
def firstItem(xs: var[]..[])
{
    return = xs[0];
}
xs = {{{1, 2, 3}, {4, 5, 6}}, {{7, 8, 9}, {10, 11, 12}}};
rs = firstItem(xs<1><1>);
```
Here `firstItem(xs<1><1>)` will return the first item of each sub-list, so `rs` is `{{1, 4}, {7, 10}}`. 

### Reviewer
@Benglin 

### FYIs
@kronz @Racel @mccrone

### CODE REVIEW

**Comment by kronz** [comment]
@ke-yu trying to wrap my head around this.  Currently, the behavior of this
xs = {{{1, 2, 3}, {4, 5, 6}}, {{7, 8, 9}, {10, 11, 12}}};
rs = firstItem(xs<1><1>);
is to return the firt nested list:
{{1,2,3},{7,8,9}}
That is, ds seems to ignore the second <1>, with users getting the same result if using  firstItem(xs<1><1>) or  firstItem(xs<1>);
So this looks like a good step in the direction of making it clearer that <> is about traversing the "levels" of the list, and the numbers are about how to match those layers.  Said another way, this is closer to your specification "Each replication guide level represents a nest loop level. Iterate each level, for arguments with same replication guide values, zip replication applies; otherwise cartesian replication applies."  
Consistency in this approach is great, so if I'm understanding this correctly, this PR LGTM!

File: src/Engine/ProtoCore/Lang/Replication/Replicator.cs [hunk 1]
```diff
@@ -139,6 +136,111 @@ public static List<ReplicationInstruction> BuildPartialReplicationInstructions(L
...
+        ///       zs  |  1  |  1  |  3  |
+        ///
+        /// This function goes through each level and calculate replication instructions.
+        /// 
+        /// replication instructions on level 0:
+        ///     Zip replication on (0, 1, 2) (i.e., zip on xs, ys, zs)
+        ///
+        /// replication instructions on level 1:
+        ///     Zip replication on (1, 2)    (i.e., zip on ys, zs)
+        ///     Cartesian replication on 0   (i.e., on xs)
```
Comment (Benglin) [inline src/Engine/ProtoCore/Lang/Replication/Replicator.cs]: Should be `Cartesian replication on argument #0`

File: src/Engine/ProtoCore/Lang/Replication/Replicator.cs [hunk 2]
```diff
@@ -139,6 +136,111 @@ public static List<ReplicationInstruction> BuildPartialReplicationInstructions(L
...
+        /// <param name="partialRepGuides"></param>
+        /// <returns></returns>
+        public static List<ReplicationInstruction> BuildPartialReplicationInstructions(List<List<ReplicationGuide>> partialRepGuides)
+        {
+            if (!partialRepGuides.Any())
+            {
+                return new List<ReplicationInstruction>();
+            }
+
+            int maxGuideLevel= partialRepGuides.Max(gs => gs.Count);
```
Comment (Benglin) [inline src/Engine/ProtoCore/Lang/Replication/Replicator.cs]: Sorry, I'm going to do some of these commenting just to make sure I get the thing right, and also help future readers.

```csharp
// Compute the maximum guide level, for example, the following would have 
// the value of maxGuideLevel equal to 3:
// 
//    foo(xs<1>, ys<1><1><2>, zs);
// 
// 
```

File: src/Engine/ProtoCore/Lang/Replication/Replicator.cs [hunk 3]
```diff
@@ -139,6 +136,111 @@ public static List<ReplicationInstruction> BuildPartialReplicationInstructions(L
...
+        ///     Zip replication on (1, 2)    (i.e., zip on ys, zs)
+        ///     Cartesian replication on 0   (i.e., on xs)
+        ///
+        /// replication instructions on level 2:
+        ///     Cartesian replication on 1   (i.e., on ys)
+        ///     Zip replication on (0, 2)    (i.e., zip on xs, zs)
+        /// </summary>
+        /// <param name="partialRepGuides"></param>
+        /// <returns></returns>
+        public static List<ReplicationInstruction> BuildPartialReplicationInstructions(List<List<ReplicationGuide>> partialRepGuides)
```
Comment (Benglin) [inline src/Engine/ProtoCore/Lang/Replication/Replicator.cs]: I'm curious as to why it is called `partialRepGuides` and not just `replicationGuides`?

File: src/Engine/ProtoCore/Lang/Replication/Replicator.cs [hunk 4]
```diff
@@ -139,6 +136,111 @@ public static List<ReplicationInstruction> BuildPartialReplicationInstructions(L
...
+            }
+
+            int maxGuideLevel= partialRepGuides.Max(gs => gs.Count);
+
+            var instructions = new List<ReplicationInstruction>();
+            for (int level = 0; level < maxGuideLevel; ++level)
+            {
+                var positions = new Dictionary<int, List<int>>();
+                var algorithms = new Dictionary<int, ZipAlgorithm>();
+                var guides = new HashSet<int>();
```
Comment (Benglin) [inline src/Engine/ProtoCore/Lang/Replication/Replicator.cs]: From my reading of the codes below, I'd think it helps clarify things if `guides` were to be renamed as `guidesOnLevel`.

File: src/Engine/ProtoCore/Lang/Replication/Replicator.cs [hunk 5]
```diff
@@ -139,6 +136,111 @@ public static List<ReplicationInstruction> BuildPartialReplicationInstructions(L
...
+                for (int i = 0; i < partialRepGuides.Count; ++i)
+                {
+                    var replicationGuides = partialRepGuides[i];
+                    if (replicationGuides == null || replicationGuides.Count <= level)
+                        continue;
+
+                    // If it is negative or 0, treat it as a stub
+                    var guide = replicationGuides[level].guideNumber;
+                    if (guide <= 0)
+                        continue;
```
Comment (Benglin) [inline src/Engine/ProtoCore/Lang/Replication/Replicator.cs]: If we rename the above `guides`, this `guide` can be renamed to `guideNumber` for better alignment.

File: src/Engine/ProtoCore/Lang/Replication/Replicator.cs [hunk 6]
```diff
@@ -139,6 +136,111 @@ public static List<ReplicationInstruction> BuildPartialReplicationInstructions(L
...
+
+                    var algorithm = replicationGuides[level].isLongest ? ZipAlgorithm.Longest : ZipAlgorithm.Shortest;
+
+                    guides.Add(guide);
+
+                    List<int> positionList = null;
```
Comment (Benglin) [inline src/Engine/ProtoCore/Lang/Replication/Replicator.cs]:
```csharp
// "positionList" contains argument index of a given "guide number" on the "guide 
// level". For example, the following call has two "guide level" (i.e. '0' and '1').
// When "guide level" is '1' (i.e. the second level in), the values of "positionList" 
// would have been { 0, 2 } (the argument indices) when "guide number" is "2":
// 
//      foo(xs<1><2>, ys<1><1>, zs<1><2>)
//                ^                   ^
// 
```

**Comment by Racel** [comment]
@ke-yu - This sounds great. Much easier to understand than the initial replication guide behavior.

**Comment by ThomasMahon** [comment]
@ke-yu I've been looking at your example and notice that:
```
firstItem(xs) 
firstItem(xs<1>) 
firstItem(xs<1><1>) 
```

All return:
` {{1, 4}, {7, 10}}.`

Can you explain?

**Comment by Racel** [comment]
@ThomasMahon - firstItem(xs), firstItem(xs<1>), and firstItem(xs<1><1>), return different results. See example below. These replication guide changes can already be seen in Dynamo 1.0.

Think about the addition of each replication guide as a way for you to access and iterate over different levels of your list.

For firstItem(xs), the function is looking at the outermost level of your list for the first item that it finds. In this case, it returns the first item as List[0].
For firstItem(xs<1>), the function is looking at the second level in. In this case, it will iterate through the second level in and return the first item of each list it finds: List[0][0] and List[1][0]
For firstItem(xs<1><1>), the function is looking at the third level in. In this case, it will iterate through the third level in and return the first item of each list it finds: List[0][0][0], List[0][1][0], List[1][0][0]. and List[1][1][0].

Take a look at the example. Hope this helps.
image

**Comment by ThomasMahon** [comment]
@Racel  I was expect to see these results. When I posted I must have written `xs:var[]` instead of `xs:var[]..[]` as I now cant reproduce the same behaviour! I'm now getting the same outputs as you. Thanks for elaborating more on replication guides

**Pull request merged**